### DISCUSSION TITLE
Error: Cannot find external module "@agoric/vat-data"

### DISCUSSION THREAD

**Discussion by Jorge-Lopes** [Agoric/agoric-sdk#7143] [Q&A] [answered]
Hello everyone 🙂 

Request:
Does anyone have a suggestion on how to debug this error:
`Cannot find external module "@agoric/vat-data"`

Context:
Our code follow the same approach of the `vaultFactory`, from `inter-protocol` package, which means that we have a contract similar to `vault.js` where `vat-data` is being imported:
`import { M, prepareExoClassKit } from '@agoric/vat-data'; `
A file similar to `vault-contract-wrapper.js` to bootstrap the testing environment, which is importing the `prepare*` function exported by the contract.
And a test file similar to `test-vault.js` that is being executed.
I am using the most recently updated master branch of agoric-sdk.

When executing a test, the following error message is printed:
```
  Rejected promise returned by test. Reason:

  TypeError {
    message: 'Failed to load module "./test/setup.js" in package "<path to "contract" directory>" 
    (1 underlying failures: Cannot find external module "@agoric/vat-data" in package 
    <path to "contract" directory>', }
 ```
Although, I verified that the imported package is linked to the correct index.js, exporting the desired methods.
And the package can be found on the node-modules folder.

A solution proposed by @dckc was to replace
`import { test } from './prepare-test-env-ava.js';` 
with
`import { test } from '@agoric/swingset-vat/tools/prepare-test-env-ava.js'; `

The error returned was:
```
Uncaught exception in test/test-rental.js

  AssertionError [ERR_ASSERTION] [ERR_ASSERTION]: The expression evaluated to a falsy value:

    assert(refs.runnerChain)

    at async Promise.all (index 0)

SES_UNHANDLED_REJECTION: (AssertionError#1)
```

Thank you very much in advance.

**Comment by tgrecojs** [comment]
@Jorge-Lopes i'm seeing now that this file from zoe just exports the test set up from swingset-vat, but FWIW, i often import it `{test}` from [@zoe/tools/prepare-test-env-ava.js](https://github.com/Agoric/agoric-sdk/blob/master/packages/zoe/tools/prepare-test-env-ava.js)

**Reply by Jorge-Lopes** [reply]
I have tried to copy the `vault-contract-wrapper.js` and `test-vault.js` to my local repository, and updated the local imports of the vault-contract-wrapper.

When running the test-vault, the return was the same as before:
```
npx ava test/test-vault.js         

  Uncaught exception in test/test-vault.js

  AssertionError [ERR_ASSERTION] [ERR_ASSERTION]: The expression evaluated to a falsy value:

    assert(refs.runnerChain)

    at async Promise.all (index 0)

SES_UNHANDLED_REJECTION: (AssertionError#1)
  ✖ test/test-vault.js exited with a non-zero exit code: 1
  ```

**Comment by dckc** [comment, marked as answer]
One clue I heard:

Is swingset doing the bundling that wants to pull this in? Is that swingset told to use `config.includeDevDependencies = true`? Is `@agoric/vat-data` included in the enclosing package's package.json under `dependencies:` or under `devDependencies:` ? Our `bundleSource` defaults to only looking at `dependencies` without that extra option.

**Reply by Jorge-Lopes** [reply]
It worked!!!
After adding `"@agoric/vat-data": "beta"` to the `dependencies` of `package.json`, the contract was able to recognise it.

Thank you very much Dan.

**Reply by Jorge-Lopes** [reply]
Although, the test setup is being imported from the local `'./prepare-test-env-ava.js'` file.
When using the:
```
import { test } from '@agoric/swingset-vat/tools/prepare-test-env-ava.js';
or
import { test } from '@agoric/zoe/tools/prepare-test-env-ava.js';
```
The AssertionError mentioned above is returned.

**Comment by dckc** [comment]
another hypothesis is that something has not been published in npm correctly. (I have only a vague notion of what this means.)
